packages/core/lib/debug/cli.js

```diff
@@ -9,33 +9,61 @@ const Codec = require("@truffle/codec");
 
 const { DebugInterpreter } = require("./interpreter");
 const { DebugCompiler } = require("./compiler");
+const { DebugExternalHandler } = require("./external");
 
 class CLIDebugger {
-  constructor(config, { compilations } = {}) {
+  constructor(config, { compilations, txHash } = {}) {
     this.config = config;
     this.compilations = compilations;
+    this.txHash = txHash;
   }
 
-  //note: txHash is optional
-  async run(txHash) {
+  async run() {
     this.config.logger.log("Starting Truffle Debugger...");
 
     // get compilations (either by shimming compiled artifacts,
     // or by doing a recompile)
-    const compilations =
-      this.compilations || (await this.getCompilations(this.config));
+    const compilations = this.compilations || (await this.getCompilations());
 
     // invoke @truffle/debugger
-    const session = await this.startDebugger(compilations, txHash);
+    const session = await this.startDebugger(compilations);
 
     // initialize prompt/breakpoints/ui logic
-    const interpreter = await this.buildInterpreter(session, txHash);
+    const interpreter = await this.buildInterpreter(session);
 
     return interpreter;
   }
 
-  async getCompilations(config) {
-    let artifacts = await DebugUtils.gatherArtifacts(config);
+  async fetchExternalSources(bugger) {
+    const fetchSpinner = ora(
+      "Getting and compiling external sources..."
+    ).start();
+    const { badAddresses, badFetchers } = await new DebugExternalHandler(
+      bugger,
+      this.config
+    ).fetch(); //note: mutates bugger!
```

Comment:
Ooh, hm.  That does seem to make more sense, but I'm not sure how workable it is.  Like, the debugger can't just call out to the handler and get the results, because a big part of the main loop is testing what addresses have defined contexts in the debugger.  So, the debugger would have to be integrated into the main loop.  But then the debugger is doing things the debugger shouldn't be doing.  You see the problem?  I don't think that structure works.

Reply:
Hm, alright!